### PR TITLE
Optparse with environment variables

### PR DESCRIPTION
**Pull request by jonascarpay**
This introduces the `Options` module, which is a wrapper around `optparse-applicative` for dealing with environment variables.
You can see the way it's used in `server/app/Main.hs`. The output looks like this:

```
$ PORT=1234 nix run . -- --help
Usage: binplz-server [-d|--program-db PATH] [-a|--app-db PATH] [-p|--port INT]

Available options:
  -d,--program-db PATH     Path to the nix program database.
                             Environment variable: NIX_PROGRAM_DB (not set)
                             Default: /nix/var/nix/profiles/per-user/root/channels/nixos/programs.sqlite
  -a,--app-db PATH         Path to the application database.
                             Environment variable: APP_DB (not set)
                             Default: appdb.sqlite
  -p,--port INT            Port to listen on.
                             Environment variable: PORT (1234)
                             Default: 8081
  -h,--help                Show this help text
```

### CODE REVIEW

File: server/app/Main.hs [hunk 1]
```diff
@@ -1,4 +1,46 @@
...
+getConfig :: IO ServerConfig
+getConfig = runOptionParser $ \envOption ->
+  ServerConfig
+    <$> envOption
+      ( Option
+          { optHelp = "Path to the nix program database.",
+            optMetaVar = "PATH",
+            optLong = "program-db",
+            optShort = 'd',
+            optEnvKey = "NIX_PROGRAM_DB",
```
Comment (kayhide) [inline server/app/Main.hs:14]: For envvar names, I suggest `<APP_PREFIX>_<THING>` style.
Unless so, it easily go hard to track which env var is consumed by which application.

For this one, `BINPLZ_PROGRAM_DB` looks better.

File: server/app/Main.hs [hunk 2]
```diff
@@ -1,4 +1,46 @@
...
+            optShow = unProgramDB
+          }
+      )
+    <*> envOption
+      ( Option
+          { optHelp = "Path to the application database.",
+            optMetaVar = "PATH",
+            optLong = "app-db",
+            optShort = 'a',
+            optEnvKey = "APP_DB",
```
Comment (kayhide) [inline server/app/Main.hs:26]: `BINPLZ_APPLICATION_DB` or `BINPLZ_APP_DB`.

I prefer long name here. This is going to be an interface to the outer space and the scope is very large.
And normally, we don't handwrite the name env vars often.

File: server/app/Main.hs [hunk 3]
```diff
@@ -1,4 +1,46 @@
...
+            optShow = unApplicationDB
+          }
+      )
+    <*> envOption
+      ( Option
+          { optHelp = "Port to listen on.",
+            optMetaVar = "INT",
+            optLong = "port",
+            optShort = 'p',
+            optEnvKey = "PORT",
```
Comment (kayhide) [inline server/app/Main.hs:38]: As for `PORT`, we have some convention and have less risk to conflict or get confused.

But I know that the convention is not so universal that it seems good to go `BINPLZ_PORT`.

**Review by kayhide — APPROVED** [review]
This looks good!
The optparse code is shorter than I expected.

I left some comments about the name of env vars.
After handling them, please feel free to merge.

Thanks!

**Comment by kayhide** [comment]
Ah, this PR is against my branch.
I thought it is against `main`.
Okay, feel freer to merge.

**Comment by jonascarpay** [comment]
@kayhide I addressed your concerns in 5d963d0935be49751365e3de24899fc213529178

**Pull request merged**